src/dfcx_scrapi/tools/evaluations.py

```diff
@@ -268,6 +269,154 @@ def run_query_and_eval(self, df: pd.DataFrame) -> pd.DataFrame:
         df = self.clean_outputs(df)
 
         return df
+    
+    def parse_interactions_from_conversation_history(self, conversation):
+        t = Tools()
+        p = Playbooks(self.agent_id)
+
+        tools_map = t.get_tools_map(self.agent_id)
+        playbooks_map = p.get_playbooks_map(self.agent_id)
+        
+        results = []
+        
+        for interaction in reversed(conversation.interactions):
+            interaction_result = {
+                "query": interaction.request.query_input.text.text,
+                "tool_calls": [],
+                "responses": []
+            }
+            
+            query_result = interaction.response.query_result
+
+            if hasattr(query_result, 'generative_info'):
+                for action in query_result.generative_info.action_tracing_info.actions:
+                    if "tool_use" in action:
```

Comment:
For each of these:
  `if tool_use...`
  `if agent_utterance...`
  `if playbook_invocation...`

We can break them into their own methods like `parse_tool_use`, `parse_agent_utterance`, etc.
This is also another opportunity to create reusable code that can live in a `core` class like `ConversationHistory`. 
There's already some similar code there so I think it would make sense to put this in there as well.